File: src/modelos.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TypeAlias


EstadoFisico: TypeAlias = dict[str, str]


def _nome_normalizado(nome: str) -> str:
    """Normaliza nomes somente para comparações e detecção de conflitos."""
    return str(nome).strip().casefold()


def _validar_nome(nome: str, descricao: str) -> None:
    if not str(nome).strip():
        raise ValueError(f"{descricao} precisa possuir um nome.")


def _verificar_repetidos(
    nomes: list[str] | tuple[str, ...],
    descricao: str,
) -> None:
    normalizados = [_nome_normalizado(nome) for nome in nomes]

    repetidos = {
        nome
        for nome in normalizados
        if normalizados.count(nome) > 1
    }

    if repetidos:
        raise ValueError(
            f"{descricao} possui nomes repetidos: "
            + ", ".join(sorted(repetidos))
            + "."
        )


@dataclass(frozen=True)
class AtuadorConfig:
    """
    Configuração de um atuador e de suas posições monitoradas.

    A ordem dos sensores representa a ordem física das posições.

    Exemplo:

        AtuadorConfig(
            nome="B",
            sensores=("b0", "b1", "b2", "b3"),
            sensor_inicial="b0",
        )

    Nesse caso:

        B+ pode avançar de b0 para b1, b2 ou b3;
        B- pode retornar de b3 para b2, b1 ou b0.
    """

    nome: str
    sensores: tuple[str, ...]
    sensor_inicial: str

    def validar(self) -> None:
        _validar_nome(
            self.nome,
            "O atuador",
        )

        if len(self.sensores) < 2:
            raise ValueError(
                f"O atuador {self.nome} precisa possuir "
                "pelo menos dois sensores."
            )

        for sensor in self.sensores:
            _validar_nome(
                sensor,
                f"Um sensor do atuador {self.nome}",
            )

        _verificar_repetidos(
            self.sensores,
            f"O atuador {self.nome}",
        )

        sensores_normalizados = {
            _nome_normalizado(sensor): sensor
            for sensor in self.sensores
        }

        if (
            _nome_normalizado(self.sensor_inicial)
            not in sensores_normalizados
        ):
            raise ValueError(
                f"O sensor inicial {self.sensor_inicial} "
                f"não pertence ao atuador {self.nome}."
            )

    def indice_sensor(self, sensor: str) -> int:
        """
        Retorna a posição ordinal de um sensor.

        A comparação é feita sem diferenciar letras maiúsculas
        e minúsculas.
        """

        procurado = _nome_normalizado(sensor)

        for indice, nome_sensor in enumerate(self.sensores):
            if _nome_normalizado(nome_sensor) == procurado:
                return indice

        raise ValueError(
            f"O sensor {sensor} não pertence ao atuador {self.nome}."
        )

    def sensor_canonico(self, sensor: str) -> str:
        """
        Retorna o nome do sensor exatamente como foi cadastrado.
        """

        return self.sensores[self.indice_sensor(sensor)]

    @property
    def quantidade_posicoes(self) -> int:
        return len(self.sensores)

    @property
    def sensor_minimo(self) -> str:
        return self.sensores[0]

    @property
    def sensor_maximo(self) -> str:
        return self.sensores[-1]


@dataclass(frozen=True)
class Movimento:
    """
    Movimento normalizado de um atuador.

    O destino é sempre explícito no modelo interno, mesmo quando
    o usuário digitou apenas A+ ou A-.

    Exemplos:

        Movimento("A", "+", "a1")
        Movimento("B", "+", "b2")
        Movimento("B", "-", "b0")
    """

    atuador: str
    sentido: str
    sensor_destino: str

    def validar(self) -> None:
        _validar_nome(
            self.atuador,
            "O movimento",
        )

        if self.sentido not in {"+", "-"}:
            raise ValueError(
                f"Sentido inválido no movimento do atuador "
                f"{self.atuador}: {self.sentido!r}."
            )

        _validar_nome(
            self.sensor_destino,
            (
                f"O sensor de destino do movimento "
                f"{self.atuador}{self.sentido}"
            ),
        )

    @property
    def saida(self) -> str:
        """
        Nome da saída física do atuador.

        Vários movimentos podem usar a mesma saída, por exemplo:

            B+ até b1
            B+ até b2
            B+ até b3

        Todos possuem a saída física B+.
        """

        return f"{self.atuador}{self.sentido}"

    @property
    def descricao(self) -> str:
        return (
            f"{self.atuador}{self.sentido} "
            f"até {self.sensor_destino}"
        )


@dataclass(frozen=True)
class EtapaSequencial:
    """
    Uma etapa da sequência.

    Uma etapa pode possuir um único movimento ou vários movimentos
    simultâneos.
    """

    movimentos: tuple[Movimento, ...]

    def validar(self) -> None:
        if not self.movimentos:
            raise ValueError(
                "Uma etapa não pode estar vazia."
            )

        for movimento in self.movimentos:
            movimento.validar()

        atuadores = [
            _nome_normalizado(movimento.atuador)
            for movimento in self.movimentos
        ]

        if len(set(atuadores)) != len(atuadores):
            raise ValueError(
                "Um mesmo atuador não pode aparecer duas vezes "
                "na mesma etapa."
            )

    @property
    def simultanea(self) -> bool:
        return len(self.movimentos) > 1

    @property
    def descricao(self) -> str:
        textos = [
            movimento.descricao
            for movimento in self.movimentos
        ]

        if self.simultanea:
            return "(" + ", ".join(textos) + ")"

        return textos[0]


@dataclass(frozen=True)
class LoopConfig:
    """
    Configuração de um trecho repetitivo da sequência.

    Os índices de início e fim são inclusivos e começam em zero.

    Exemplo:

        LoopConfig(
            inicio=2,
            fim=5,
            sensor="e",
            repetir_quando=0,
        )

    Significa que as etapas 2, 3, 4 e 5 são repetidas enquanto
    e = 0. Quando e = 1, a sequência continua após a etapa 5.
    """

    inicio: int
    fim: int
    sensor: str
    repetir_quando: int = 0

    @property
    def sair_quando(self) -> int:
        return 1 - self.repetir_quando

    def validar(self, quantidade_etapas: int) -> None:
        if quantidade_etapas <= 0:
            raise ValueError(
                "Não é possível criar um loop sem etapas."
            )

        if self.inicio < 0 or self.inicio >= quantidade_etapas:
            raise ValueError(
                "O início do loop é inválido."
            )

        if self.fim < self.inicio or self.fim >= quantidade_etapas:
            raise ValueError(
                "O final do loop é inválido."
            )

        if self.repetir_quando not in {0, 1}:
            raise ValueError(
                "A condição de repetição precisa ser 0 ou 1."
            )

        _validar_nome(
            self.sensor,
            "O sensor de decisão do loop",
        )

    @property
    def quantidade_etapas(self) -> int:
        return self.fim - self.inicio + 1

    @property
    def descricao(self) -> str:
        return (
            f"etapas {self.inicio + 1} a {self.fim + 1}, "
            f"repetir enquanto {self.sensor} = "
            f"{self.repetir_quando}"
        )


@dataclass
class ProjetoSequencial:
    """
    Modelo completo de entrada do solucionador.

    Ele será utilizado tanto pela entrada textual quanto pelo futuro
    editor visual. O motor deverá receber este modelo já validado.
    """

    atuadores: dict[str, AtuadorConfig]
    etapas: list[EtapaSequencial]
    loops: list[LoopConfig] = field(default_factory=list)
    sinal_partida: str = "S"
    entradas_externas: list[str] = field(default_factory=list)

    def validar(self) -> None:
        self._validar_estrutura_basica()
        self._validar_atuadores()
        self._validar_entradas_externas()
        self._validar_conflitos_de_nomes()

        estados = self._validar_e_simular_etapas()

        self._validar_loops(
            estados=estados,
        )

    def _validar_estrutura_basica(self) -> None:
        if not self.atuadores:
            raise ValueError(
                "É necessário cadastrar pelo menos um atuador."
            )

        if not self.etapas:
            raise ValueError(
                "É necessário cadastrar pelo menos uma etapa."
            )

        _validar_nome(
            self.sinal_partida,
            "O sinal de partida",
        )

    def _validar_atuadores(self) -> None:
        nomes_atuadores: list[str] = []

        for chave, atuador in self.atuadores.items():
            atuador.validar()

            if _nome_normalizado(chave) != _nome_normalizado(
                atuador.nome
            ):
                raise ValueError(
                    f"A chave {chave!r} não corresponde "
                    f"ao atuador {atuador.nome!r}."
                )

            nomes_atuadores.append(atuador.nome)

        _verificar_repetidos(
            nomes_atuadores,
            "O cadastro de atuadores",
        )

    def _validar_entradas_externas(self) -> None:
        for entrada in self.entradas_externas:
            _validar_nome(
                entrada,
                "Uma entrada externa",
            )

        _verificar_repetidos(
            self.entradas_externas,
            "O cadastro de entradas externas",
        )

    def _validar_conflitos_de_nomes(self) -> None:
        """
        Impede que dois elementos diferentes utilizem o mesmo nome.

        A comparação ignora diferenças entre maiúsculas e minúsculas.
        """

        ocupados: dict[str, str] = {}

        def registrar(
            nome: str,
            descricao: str,
        ) -> None:
            chave = _nome_normalizado(nome)

            anterior = ocupados.get(chave)

            if anterior is not None:
                raise ValueError(
                    f"O nome {nome!r} está sendo utilizado por "
                    f"{descricao} e também por {anterior}."
                )

            ocupados[chave] = descricao

        for atuador in self.atuadores.values():
            registrar(
                atuador.nome,
                f"o atuador {atuador.nome}",
            )

            for sensor in atuador.sensores:
                registrar(
                    sensor,
                    (
                        f"o sensor {sensor} do atuador "
                        f"{atuador.nome}"
                    ),
                )

        for entrada in self.entradas_externas:
            registrar(
                entrada,
                f"a entrada externa {entrada}",
            )

        registrar(
            self.sinal_partida,
            f"o sinal de partida {self.sinal_partida}",
        )

    def _obter_atuador(
        self,
        nome: str,
    ) -> AtuadorConfig:
        procurado = _nome_normalizado(nome)

        for chave, configuracao in self.atuadores.items():
            if (
                _nome_normalizado(chave) == procurado
                or _nome_normalizado(configuracao.nome) == procurado
            ):
                return configuracao

        raise ValueError(
            f"O atuador {nome} não foi cadastrado."
        )

    def _validar_e_simular_etapas(
        self,
    ) -> list[EstadoFisico]:
        """
        Valida os movimentos e produz os estados lineares.

        A lista retornada possui:

            estados[0] = estado inicial
            estados[1] = estado após a etapa 1
            estados[2] = estado após a etapa 2
            ...

        Portanto, o estado anterior à etapa de índice i é estados[i],
        e o estado posterior é estados[i + 1].
        """

        estado_atual: EstadoFisico = {
            atuador.nome: atuador.sensor_canonico(
                atuador.sensor_inicial
            )
            for atuador in self.atuadores.values()
        }

        estados: list[EstadoFisico] = [
            dict(estado_atual)
        ]

        for indice_etapa, etapa in enumerate(
            self.etapas,
            start=1,
        ):
            etapa.validar()

            proximo_estado = dict(estado_atual)

            for movimento in etapa.movimentos:
                configuracao = self._obter_atuador(
                    movimento.atuador
                )

                sensor_atual = estado_atual[
                    configuracao.nome
                ]

                indice_atual = configuracao.indice_sensor(
                    sensor_atual
                )

                indice_destino = configuracao.indice_sensor(
                    movimento.sensor_destino
                )

                if indice_destino == indice_atual:
                    raise ValueError(
                        f"Etapa {indice_etapa}: o atuador "
                        f"{configuracao.nome} já está no sensor "
                        f"{movimento.sensor_destino}."
                    )

                if (
                    movimento.sentido == "+"
                    and indice_destino < indice_atual
                ):
                    raise ValueError(
                        f"Etapa {indice_etapa}: o movimento "
                        f"{movimento.saida} não pode ir de "
                        f"{sensor_atual} para "
                        f"{movimento.sensor_destino}. "
                        "O sentido positivo deve avançar na ordem "
                        "dos sensores."
                    )

                if (
                    movimento.sentido == "-"
                    and indice_destino > indice_atual
                ):
                    raise ValueError(
                        f"Etapa {indice_etapa}: o movimento "
                        f"{movimento.saida} não pode ir de "
                        f"{sensor_atual} para "
                        f"{movimento.sensor_destino}. "
                        "O sentido negativo deve retornar na ordem "
                        "dos sensores."
                    )

                proximo_estado[
                    configuracao.nome
                ] = configuracao.sensor_canonico(
                    movimento.sensor_destino
                )

            estado_atual = proximo_estado
            estados.append(
                dict(estado_atual)
            )

        return estados

    def _validar_loops(
        self,
        estados: list[EstadoFisico],
    ) -> None:
        entradas_normalizadas = {
            _nome_normalizado(entrada)
            for entrada in self.entradas_externas
        }

        for loop in self.loops:
            loop.validar(
                len(self.etapas)
            )

            if (
                _nome_normalizado(loop.sensor)
                not in entradas_normalizadas
            ):
                raise ValueError(
                    f"O sensor externo {loop.sensor} utilizado "
                    "no loop não foi cadastrado."
                )

        self._validar_sobreposicao_loops()

        for loop in self.loops:
            estado_entrada = estados[
                loop.inicio
            ]

            estado_saida = estados[
                loop.fim + 1
            ]

            if estado_entrada != estado_saida:
                diferencas: list[str] = []

                for atuador in self.atuadores.values():
                    antes = estado_entrada[
                        atuador.nome
                    ]

                    depois = estado_saida[
                        atuador.nome
                    ]

                    if antes != depois:
                        diferencas.append(
                            f"{atuador.nome}: "
                            f"{antes} → {depois}"
                        )

                raise ValueError(
                    "O trecho configurado como loop não retorna "
                    "ao estado físico necessário para uma nova "
                    "repetição. Diferenças encontradas: "
                    + ", ".join(diferencas)
                    + "."
                )

    def _validar_sobreposicao_loops(self) -> None:
        """
        Nesta primeira versão, loops podem existir em quantidade
        maior que um, mas não podem ser aninhados nem sobrepostos.
        """

        ordenados = sorted(
            self.loops,
            key=lambda loop: (
                loop.inicio,
                loop.fim,
            ),
        )

        for anterior, atual in zip(
            ordenados,
            ordenados[1:],
        ):
            if atual.inicio <= anterior.fim:
                raise ValueError(
                    "Os loops configurados entre as etapas "
                    f"{anterior.inicio + 1}–{anterior.fim + 1} "
                    f"e {atual.inicio + 1}–{atual.fim + 1} "
                    "estão sobrepostos. Loops sobrepostos ou "
                    "aninhados ainda não são permitidos."
                )

    def estado_inicial(self) -> EstadoFisico:
        """
        Retorna o estado inicial usando os sensores cadastrados.
        """

        return {
            atuador.nome: atuador.sensor_canonico(
                atuador.sensor_inicial
            )
            for atuador in self.atuadores.values()
        }

    def simular_estados(self) -> list[EstadoFisico]:
        """
        Valida o projeto e retorna todos os estados lineares.
        """

        self.validar()

        return self._validar_e_simular_etapas()

    def sensores_fisicos(self) -> tuple[str, ...]:
        """
        Retorna todos os sensores físicos na ordem de cadastro.
        """

        return tuple(
            sensor
            for atuador in self.atuadores.values()
            for sensor in atuador.sensores
        )

    def obter_loop_da_etapa(
        self,
        indice_etapa: int,
    ) -> LoopConfig | None:
        """
        Retorna o loop que contém uma etapa, quando existir.
        """

        for loop in self.loops:
            if loop.inicio <= indice_etapa <= loop.fim:
                return loop

        return None

File: mobile/lib/motor/modelos.dart
/// Erro de validação apresentado ao usuário quando a sequência é inválida.
final class EntradaInvalidaException implements Exception {
  const EntradaInvalidaException(this.mensagem);

  final String mensagem;

  @override
  String toString() => mensagem;
}

String _nomeNormalizado(String nome) => nome.trim().toLowerCase();

void _validarNome(String nome, String descricao) {
  if (nome.trim().isEmpty) {
    throw EntradaInvalidaException('$descricao precisa possuir um nome.');
  }
}

void _verificarRepetidos(Iterable<String> nomes, String descricao) {
  final normalizados = nomes.map(_nomeNormalizado).toList(growable: false);
  final repetidos = normalizados
      .where(
        (nome) => normalizados.where((candidato) => candidato == nome).length > 1,
      )
      .toSet()
      .toList()
    ..sort();

  if (repetidos.isNotEmpty) {
    throw EntradaInvalidaException(
      '$descricao possui nomes repetidos: ${repetidos.join(', ')}.',
    );
  }
}

/// Configuração de um atuador e de suas posições monitoradas.
final class AtuadorConfig {
  AtuadorConfig({
    required this.nome,
    required List<String> sensores,
    required this.sensorInicial,
  }) : sensores = List.unmodifiable(sensores);

  final String nome;
  final List<String> sensores;
  final String sensorInicial;

  void validar() {
    _validarNome(nome, 'O atuador');

    if (sensores.length < 2) {
      throw EntradaInvalidaException(
        'O atuador $nome precisa possuir pelo menos dois sensores.',
      );
    }

    for (final sensor in sensores) {
      _validarNome(sensor, 'Um sensor do atuador $nome');
    }

    _verificarRepetidos(sensores, 'O atuador $nome');

    if (!sensores.any(
      (sensor) => _nomeNormalizado(sensor) == _nomeNormalizado(sensorInicial),
    )) {
      throw EntradaInvalidaException(
        'O sensor inicial $sensorInicial não pertence ao atuador $nome.',
      );
    }
  }

  int indiceSensor(String sensor) {
    final procurado = _nomeNormalizado(sensor);

    for (var indice = 0; indice < sensores.length; indice++) {
      if (_nomeNormalizado(sensores[indice]) == procurado) {
        return indice;
      }
    }

    throw EntradaInvalidaException(
      'O sensor $sensor não pertence ao atuador $nome.',
    );
  }

  String sensorCanonico(String sensor) => sensores[indiceSensor(sensor)];

  int get quantidadePosicoes => sensores.length;

  String get sensorMinimo => sensores.first;

  String get sensorMaximo => sensores.last;
}

/// Movimento normalizado de um atuador, sempre com destino explícito.
final class Movimento {
  const Movimento({
    required this.atuador,
    required this.sentido,
    required this.sensorDestino,
  });

  final String atuador;
  final String sentido;
  final String sensorDestino;

  void validar() {
    _validarNome(atuador, 'O movimento');

    if (sentido != '+' && sentido != '-') {
      throw EntradaInvalidaException(
        "Sentido inválido no movimento do atuador $atuador: '$sentido'.",
      );
    }

    _validarNome(
      sensorDestino,
      'O sensor de destino do movimento $atuador$sentido',
    );
  }

  String get saida => '$atuador$sentido';

  String get descricao => '$saida até $sensorDestino';
}

/// Uma etapa com um ou mais movimentos simultâneos.
final class EtapaSequencial {
  EtapaSequencial({required List<Movimento> movimentos})
    : movimentos = List.unmodifiable(movimentos);

  final List<Movimento> movimentos;

  void validar() {
    if (movimentos.isEmpty) {
      throw const EntradaInvalidaException('Uma etapa não pode estar vazia.');
    }

    for (final movimento in movimentos) {
      movimento.validar();
    }

    final atuadores = movimentos
        .map((movimento) => _nomeNormalizado(movimento.atuador))
        .toList(growable: false);

    if (atuadores.toSet().length != atuadores.length) {
      throw const EntradaInvalidaException(
        'Um mesmo atuador não pode aparecer duas vezes na mesma etapa.',
      );
    }
  }

  bool get simultanea => movimentos.length > 1;

  String get descricao {
    final textos = movimentos.map((movimento) => movimento.descricao).toList();
    return simultanea ? '(${textos.join(', ')})' : textos.single;
  }
}

/// Configuração de um trecho repetitivo da sequência.
final class LoopConfig {
  const LoopConfig({
    required this.inicio,
    required this.fim,
    required this.sensor,
    this.repetirQuando = 0,
  });

  final int inicio;
  final int fim;
  final String sensor;
  final int repetirQuando;

  int get sairQuando => 1 - repetirQuando;

  int get quantidadeEtapas => fim - inicio + 1;

  String get descricao =>
      'etapas ${inicio + 1} a ${fim + 1}, '
      'repetir enquanto $sensor = $repetirQuando';

  void validar(int quantidadeTotalEtapas) {
    if (quantidadeTotalEtapas <= 0) {
      throw const EntradaInvalidaException(
        'Não é possível criar um loop sem etapas.',
      );
    }

    if (inicio < 0 || inicio >= quantidadeTotalEtapas) {
      throw const EntradaInvalidaException('O início do loop é inválido.');
    }

    if (fim < inicio || fim >= quantidadeTotalEtapas) {
      throw const EntradaInvalidaException('O final do loop é inválido.');
    }

    if (repetirQuando != 0 && repetirQuando != 1) {
      throw const EntradaInvalidaException(
        'A condição de repetição precisa ser 0 ou 1.',
      );
    }

    _validarNome(sensor, 'O sensor de decisão do loop');
  }
}

/// Modelo completo e validado de entrada do solucionador.
final class ProjetoSequencial {
  ProjetoSequencial({
    required Map<String, AtuadorConfig> atuadores,
    required List<EtapaSequencial> etapas,
    List<LoopConfig> loops = const [],
    this.sinalPartida = 'S',
    List<String> entradasExternas = const [],
  }) : atuadores = Map.unmodifiable(atuadores),
       etapas = List.unmodifiable(etapas),
       loops = List.unmodifiable(loops),
       entradasExternas = List.unmodifiable(entradasExternas);

  final Map<String, AtuadorConfig> atuadores;
  final List<EtapaSequencial> etapas;
  final List<LoopConfig> loops;
  final String sinalPartida;
  final List<String> entradasExternas;

  void validar() {
    _validarEstruturaBasica();
    _validarAtuadores();
    _validarEntradasExternas();
    _validarConflitosDeNomes();
    final estados = _validarESimularEtapas();
    _validarLoops(estados);
  }

  void _validarEstruturaBasica() {
    if (atuadores.isEmpty) {
      throw const EntradaInvalidaException(
        'É necessário cadastrar pelo menos um atuador.',
      );
    }

    if (etapas.isEmpty) {
      throw const EntradaInvalidaException(
        'É necessário cadastrar pelo menos uma etapa.',
      );
    }

    _validarNome(sinalPartida, 'O sinal de partida');
  }

  void _validarAtuadores() {
    final nomesAtuadores = <String>[];

    for (final entrada in atuadores.entries) {
      entrada.value.validar();

      if (_nomeNormalizado(entrada.key) !=
          _nomeNormalizado(entrada.value.nome)) {
        throw EntradaInvalidaException(
          "A chave '${entrada.key}' não corresponde ao atuador "
          "'${entrada.value.nome}'.",
        );
      }

      nomesAtuadores.add(entrada.value.nome);
    }

    _verificarRepetidos(nomesAtuadores, 'O cadastro de atuadores');
  }

  void _validarEntradasExternas() {
    for (final entrada in entradasExternas) {
      _validarNome(entrada, 'Uma entrada externa');
    }
    _verificarRepetidos(entradasExternas, 'O cadastro de entradas externas');
  }

  void _validarConflitosDeNomes() {
    final ocupados = <String, String>{};

    void registrar(String nome, String descricao) {
      final chave = _nomeNormalizado(nome);
      final anterior = ocupados[chave];

      if (anterior != null) {
        throw EntradaInvalidaException(
          "O nome '$nome' está sendo utilizado por $descricao "
          'e também por $anterior.',
        );
      }

      ocupados[chave] = descricao;
    }

    for (final atuador in atuadores.values) {
      registrar(atuador.nome, 'o atuador ${atuador.nome}');
      for (final sensor in atuador.sensores) {
        registrar(sensor, 'o sensor $sensor do atuador ${atuador.nome}');
      }
    }

    for (final entrada in entradasExternas) {
      registrar(entrada, 'a entrada externa $entrada');
    }

    registrar(sinalPartida, 'o sinal de partida $sinalPartida');
  }

  AtuadorConfig _obterAtuador(String nome) {
    final procurado = _nomeNormalizado(nome);

    for (final entrada in atuadores.entries) {
      if (_nomeNormalizado(entrada.key) == procurado ||
          _nomeNormalizado(entrada.value.nome) == procurado) {
        return entrada.value;
      }
    }

    throw EntradaInvalidaException('O atuador $nome não foi cadastrado.');
  }

  List<Map<String, String>> _validarESimularEtapas() {
    var estadoAtual = <String, String>{
      for (final atuador in atuadores.values)
        atuador.nome: atuador.sensorCanonico(atuador.sensorInicial),
    };
    final estados = <Map<String, String>>[Map.of(estadoAtual)];

    for (var indiceEtapa = 0; indiceEtapa < etapas.length; indiceEtapa++) {
      final etapa = etapas[indiceEtapa];
      etapa.validar();
      final proximoEstado = Map<String, String>.of(estadoAtual);

      for (final movimento in etapa.movimentos) {
        final configuracao = _obterAtuador(movimento.atuador);
        final sensorAtual = estadoAtual[configuracao.nome]!;
        final indiceAtual = configuracao.indiceSensor(sensorAtual);
        final indiceDestino = configuracao.indiceSensor(
          movimento.sensorDestino,
        );

        if (indiceDestino == indiceAtual) {
          throw EntradaInvalidaException(
            'Etapa ${indiceEtapa + 1}: o atuador ${configuracao.nome} '
            'já está no sensor ${movimento.sensorDestino}.',
          );
        }

        if (movimento.sentido == '+' && indiceDestino < indiceAtual) {
          throw EntradaInvalidaException(
            'Etapa ${indiceEtapa + 1}: o movimento ${movimento.saida} '
            'não pode ir de $sensorAtual para ${movimento.sensorDestino}. '
            'O sentido positivo deve avançar na ordem dos sensores.',
          );
        }

        if (movimento.sentido == '-' && indiceDestino > indiceAtual) {
          throw EntradaInvalidaException(
            'Etapa ${indiceEtapa + 1}: o movimento ${movimento.saida} '
            'não pode ir de $sensorAtual para ${movimento.sensorDestino}. '
            'O sentido negativo deve retornar na ordem dos sensores.',
          );
        }

        proximoEstado[configuracao.nome] = configuracao.sensorCanonico(
          movimento.sensorDestino,
        );
      }

      estadoAtual = proximoEstado;
      estados.add(Map.of(estadoAtual));
    }

    return estados;
  }

  void _validarLoops(List<Map<String, String>> estados) {
    final entradasNormalizadas = entradasExternas.map(_nomeNormalizado).toSet();

    for (final loop in loops) {
      loop.validar(etapas.length);
      if (!entradasNormalizadas.contains(_nomeNormalizado(loop.sensor))) {
        throw EntradaInvalidaException(
          'O sensor externo ${loop.sensor} utilizado no loop não foi '
          'cadastrado.',
        );
      }
    }

    _validarSobreposicaoLoops();

    for (final loop in loops) {
      final estadoEntrada = estados[loop.inicio];
      final estadoSaida = estados[loop.fim + 1];

      if (!_mapasIguais(estadoEntrada, estadoSaida)) {
        final diferencas = <String>[];
        for (final atuador in atuadores.values) {
          final antes = estadoEntrada[atuador.nome];
          final depois = estadoSaida[atuador.nome];
          if (antes != depois) {
            diferencas.add('${atuador.nome}: $antes → $depois');
          }
        }

        throw EntradaInvalidaException(
          'O trecho configurado como loop não retorna ao estado físico '
          'necessário para uma nova repetição. Diferenças encontradas: '
          '${diferencas.join(', ')}.',
        );
      }
    }
  }

  void _validarSobreposicaoLoops() {
    final ordenados = List<LoopConfig>.of(loops)
      ..sort((a, b) {
        final porInicio = a.inicio.compareTo(b.inicio);
        return porInicio != 0 ? porInicio : a.fim.compareTo(b.fim);
      });

    for (var indice = 1; indice < ordenados.length; indice++) {
      final anterior = ordenados[indice - 1];
      final atual = ordenados[indice];
      if (atual.inicio <= anterior.fim) {
        throw EntradaInvalidaException(
          'Os loops configurados entre as etapas '
          '${anterior.inicio + 1}–${anterior.fim + 1} e '
          '${atual.inicio + 1}–${atual.fim + 1} estão sobrepostos. '
          'Loops sobrepostos ou aninhados ainda não são permitidos.',
        );
      }
    }
  }

  Map<String, String> estadoInicial() => <String, String>{
    for (final atuador in atuadores.values)
      atuador.nome: atuador.sensorCanonico(atuador.sensorInicial),
  };

  List<Map<String, String>> simularEstados() {
    validar();
    return _validarESimularEtapas();
  }

  List<String> sensoresFisicos() => <String>[
    for (final atuador in atuadores.values) ...atuador.sensores,
  ];

  LoopConfig? obterLoopDaEtapa(int indiceEtapa) {
    for (final loop in loops) {
      if (loop.inicio <= indiceEtapa && indiceEtapa <= loop.fim) {
        return loop;
      }
    }
    return null;
  }
}

bool _mapasIguais(Map<String, String> primeiro, Map<String, String> segundo) {
  if (primeiro.length != segundo.length) {
    return false;
  }

  for (final entrada in primeiro.entries) {
    if (segundo[entrada.key] != entrada.value) {
      return false;
    }
  }

  return true;
}
